src/Bicep.Core/TypeSystem/DeployTimeConstantViolationVisitor.cs

```diff
@@ -0,0 +1,100 @@
+        {
+            this.DeployTimeConstantContainer = deployTimeConstantContainer;
+            this.SemanticModel = semanticModel;
+            this.DiagnosticWriter = diagnosticWriter;
+        }
+
+        protected SyntaxBase DeployTimeConstantContainer { get; }
+
+        protected SemanticModel SemanticModel { get; }
+
+        protected IDiagnosticWriter DiagnosticWriter { get; }
+
+        protected void FlagDeployTimeConstantViolation(SyntaxBase errorSyntax, DeclaredSymbol? accessedSymbol = null, ObjectType? accessedObjectType = null, IEnumerable<string>? variableDependencyChain = null)
+        {
+            var accessedSymbolName = accessedSymbol?.Name;
+            var accessiblePropertyNames = GetAccessiblePropertyNames(accessedSymbol, accessedObjectType);
+
+            var diagnosticBuilder = DiagnosticBuilder.ForPosition(errorSyntax);
+            var diagnostic = this.DeployTimeConstantContainer switch
+            {
+                ObjectPropertySyntax propertySyntax when propertySyntax.TryGetKeyText() is { } propertyName => diagnosticBuilder.RuntimeValueNotAllowedInProperty(propertyName, accessedSymbolName, accessiblePropertyNames, variableDependencyChain),
+                IfConditionSyntax => diagnosticBuilder.RuntimeValueNotAllowedInIfConditionExpression(accessedSymbolName, accessiblePropertyNames, variableDependencyChain),
+
+                // Corner case: the runtime value is in the for-body of a variable declaration.
+                ForSyntax forSyntax when ErrorSyntaxInForBodyOfVariable(forSyntax, errorSyntax) is string variableName =>
+                    diagnosticBuilder.RuntimeValueNotAllowedInVariableForBody(variableName, accessedSymbolName, accessiblePropertyNames, variableDependencyChain),
+
+                ForSyntax => diagnosticBuilder.RuntimeValueNotAllowedInForExpression(accessedSymbolName, accessiblePropertyNames, variableDependencyChain),
+                FunctionCallSyntaxBase functionCallSyntaxBase => diagnosticBuilder.RuntimeValueNotAllowedInRunTimeFunctionArguments(functionCallSyntaxBase.Name.IdentifierName, accessedSymbolName, accessiblePropertyNames, variableDependencyChain),
+                _ => throw new ArgumentOutOfRangeException(nameof(this.DeployTimeConstantContainer), "Expected an ObjectPropertySyntax with a propertyName, a IfConditionSyntax, a ForSyntax, or a FunctionCallSyntaxBase."),
+            };
+
+            this.DiagnosticWriter.Write(diagnostic);
+        }
+
+        private string? ErrorSyntaxInForBodyOfVariable(ForSyntax forSyntax, SyntaxBase errorSyntax) =>
+            this.SemanticModel.Binder.GetParent(forSyntax) is VariableDeclarationSyntax variableDeclarationSyntax &&
+            TextSpan.AreOverlapping(errorSyntax, forSyntax.Body)
+                ? variableDeclarationSyntax.Name.IdentifierName
+                : null;
+
+        protected (DeclaredSymbol?, ObjectType?) TryExtractResourceOrModuleSymbolAndBodyType(SyntaxBase syntax, bool isCollection = false)
```

Comment:
>false [](http://example.com/codeflow?start=132&length=5)

Should we drop the default value here to force every caller to pass it in explicitly? (Could avoid mistakes.)

Reply:
Makes sense. Removed the default value.